File: codes/HellDream/74.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int rows = matrix.length;
        if(matrix==null||rows==0) return false;
        int cols = matrix[0].length;
        if(cols==0) return false;
        int start = 0, end = rows-1;
        while(start+1<end){
            int mid = start+(end-start)/2;
            if(matrix[mid][0]==target) return true;
            if(matrix[mid][0]<target){
                start = mid;
            }else{
                end = mid;
            }
        }
        int left = 0, right = cols-1;
        int beginRow = start;
        if(matrix[end][0]<=target){
            beginRow = end;
        }
        while(left+1<right){
            int mid = left+(right-left)/2;
            if(matrix[beginRow][mid]==target) return true;
            if(matrix[beginRow][mid]<target){
                left = mid;
            }else{
                right = mid;
            }
        }
        if(matrix[beginRow][left]==target) return true;
        if(matrix[beginRow][right]==target) return true;
        
        return false;
    }
}

File: codes/Garnetwzy/74.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int row = matrix.size();
        if(row == 0)
            return false;
        int colomn = matrix[0].size();
        if(colomn == 0)
            return false;
        int s = 0, e = row-1;
        while(s < e) {
            int mid = s + (e - s ) / 2;
            if(matrix[mid][0] < target) {
                s = mid + 1;
            }else {
                e = mid;
            }
        }
        int targetRow;
        if(matrix[s][0] == target) {
            return true;
        } else if (matrix[s][0] > target) {
            if(s == 0) {
                return false;
            }else {
                targetRow = s - 1;
            }
        } else {
            targetRow = s;
        }
        s = 0;
        e = colomn-1;
        while(s <= e) {
            int mid = s + (e - s) / 2;
            if(matrix[targetRow][mid] == target)
                return true;
            if(matrix[targetRow][mid] > target) {
                e = mid - 1;
            }else {
                s = mid + 1;
            }
        }
        return false;
    }
};